File: src/csharp/ef-api/Data/ResultsRepository.cs
using Microsoft.EntityFrameworkCore;

namespace EFExample.Data;

public record RunnerRaceResult(
  string RunnerName,
  string RaceName,
  int Position,
  TimeSpan Time,
  DateTime RaceDate
);

public class ResultsRepository(
  Database db // 👈 Injected via DI
) {
  public async Task<IEnumerable<RunnerRaceResult>> Top10FinishesByRunner(string email)
    => (await db.Runners
      .Where(r => r.Email == email)
      .SelectMany(r => r.RaceResults!.Where(
        finish => finish.Position <= 10)
      )
      // ✨ Notice how everything is fully typed downstack
      .Select(finish => new {
          RunnerName = finish.Runner.Name,
          RaceName = finish.Race.Name,
          finish.Position,
          finish.Time,
          RaceDate = finish.Race.Date
        }
      )
      .OrderBy(r => r.Position)
      .ToListAsync())
      .Select(r => new RunnerRaceResult(
        r.RunnerName,
        r.RaceName,
        r.Position,
        r.Time,
        r.RaceDate
      ));
}

File: src/typescript/prisma-api/src/results-repository.ts
import { Injectable } from "@nestjs/common";
import { PrismaService } from "./prisma.service";

@Injectable()
export class ResultsRepository {
  constructor(private readonly prismaService: PrismaService) {}

  async top10FinishesByRunner(email: string) : Promise<RaceResultDto[]> {
    return (await this.prismaService.raceResult.findMany({
      select:{
        runner: { select: { name: true } },
        race: { select: { name: true, date: true } },
        position: true,
        time: true
      },
      where: {
        runner: { email: 'ada@example.org' },
        position: { lte: 10 }
      },
      orderBy: { position: 'asc' }
    })).map(r => ({
      runnerName: r.runner.name,
      raceName: r.race.name,
      position: r.position,
      time: r.time,
      raceDate: r.race.date
    }))
  }

  async runnerResults(email: string) {
    return await this.prismaService.runner.findFirst({
      where: {
        email: 'ada@example.org'
      },
      include: {
        races: { include: { race: true } }
      }
    })
  }
}

export type RaceResultDto = {
  runnerName: string,
  raceName: string,
  position: number,
  time: number,
  raceDate: Date
}
